webapp/src/projects.tsx

```diff
@@ -416,6 +414,98 @@ export class ProjectsMenu extends data.Component<ISettingsProps, {}> {
     }
 }
 
+interface HeroBannerState {
+    cardIndex?: number;
+}
+
+const HERO_BANNER_DELAY = 10000;
+class HeroBanner extends data.Component<ISettingsProps, HeroBannerState> {
+    private prevGalleries: pxt.CodeCard[] = [];
+    private carouselInterval: any = undefined;
+
+    constructor(props: ProjectsCarouselProps) {
+        super(props)
+        this.handleRefreshCard = this.handleRefreshCard.bind(this);
+        this.handleCardClick = this.handleCardClick.bind(this);
+        this.state = {
+        }
+    }
+
+    private handleRefreshCard() {
+        pxt.log(`next hero carousel`)
+        const cardIndex = this.state.cardIndex || 0;
+        this.setState({ cardIndex: (cardIndex + 1) % this.prevGalleries.length })
+    }
+
+    private handleCardClick() {
+        const card = this.state.cardIndex !== undefined && this.prevGalleries[this.state.cardIndex]
+        if (card)
+            pxt.tickEvent("hero.card.click", { card: card.name })
+    }
+
+    private startRefresh() {
+        if (!this.carouselInterval && this.prevGalleries.length) {
+            pxt.log(`start refreshing hero carousel`)
+            this.carouselInterval = setInterval(this.handleRefreshCard, HERO_BANNER_DELAY);
+            if (this.state.cardIndex === undefined)
+                setTimeout(this.handleRefreshCard, 100); // show first item immediately
+        }
+    }
+
+    componentWillMount() {
+        this.startRefresh();
+    }
+
+    componentWillUnmount() {
+        if (this.carouselInterval) {
+            pxt.log(`stopping hero carousel`)
+            clearInterval(this.carouselInterval)
+            this.carouselInterval = undefined;
+        }
+    }
+
+    fetchGallery(): pxt.CodeCard[] {
+        const targetTheme = pxt.appTarget.appTheme;
+        const path = targetTheme.homeScreenHeroGallery;
+        if (!path) {
+            return this.prevGalleries = [];
+        }
+
+        // fetch gallery
+        let res = this.getData(`gallery:${encodeURIComponent(path)}`) as pxt.gallery.Gallery[];
```

Comment:
Just wondering: how is this a synchronous call?

Reply:
``gallery`` is a async virtual api; so getData will return undefined until it returns.